llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -1790,6 +1790,159 @@ static Instruction *foldSelectICmpEq(SelectInst &SI, ICmpInst *ICI,
   return nullptr;
 }
 
+// ICmpInst of SelectInst is not included in the calculation of KnownBits
+// so we are missing the opportunity to optimize the Value of the True or
+// False Condition via ICmpInst with KnownBits.
```

Comment:
What about: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/ValueTracking.cpp#L1046?

Reply:
IMO, same way to use  computeKnownBitsFromCond, but computeKnownBitsFromOperator would compute known bits of SelectInst. I use Cond from SelectInst to estimate the KnownBits of the variables that make up Cond. The key is that if you can use it to estimate the known bits of Trueval, there is a difference in performing the optimization.

Reply:
Oh I see what you mean. This seems very similar to #88298, I think generally this code would fit better as an extension to foldSelectValueEquivilence.

Reply:
Still feel like the code fits better in `foldSelectValueEquivilence`

Reply:
@goldsteinn 
I changed it to call inside foldSelectValueEquivalence.


The order is changed so that the optimizations I added are performed first, preventing the previously performed optimizations from being performed. If this happens, should I change the order of the other optimization?